### ISSUE TITLE
Custom / site-restricted search, like Rollyo or Bing Custom Search

### ISSUE THREAD

**Issue by wcerfgba** [searx/searx#2952]
**Is your feature request related to a problem? Please describe.**
Over time I come to favour particular websites for certain information and I want to search those sources first. For example, if I wanted a burrito recipe, I could just search for “burrito recipe” on a standard search engine, but over the years I’ve discovered a few recipe websites which I really like, and so I prefer to search them first as my preferred sources, before searching the entire Web.

**Describe the solution you'd like**
Ability to configure a list of preferred sites which are searched first, by restricting the search query via the `site:` parameter. In Google, Bing and DuckDuckGo, you can restrict a search to a single website with `burrito site:https://naturallyella.com/`. I want to restrict to a list of sites and then combine the results from all of these queries.

**Describe alternatives you've considered**
1. Implement new engine for each site. Not really appropriate because many engines can search a single site.
2. Search each site manually. Does not scale to using many websites.

**Additional context**
I've hacked this on to Searx myself on my fork at <https://github.com/wcerfgba/searx> , but it's not ready for a pull request yet. I wanted to share the idea with the world to see if people think this is worth me tidying up and submitting a proper PR for the main project. If so, notes on project code style and how to test are welcome! :smile:  See also my blog post for a tiny bit more exposition on why I want this: <https://jpreston.xyz/custom-search-engine.html>

**Comment by kvch** [comment]
I have seen your post on HN. :) How is this feature is different from the existing categories? You are free to create your own categories and add engines to that category. Then you can run `!my-custom-category` and search in those engines. You implemented "priority engines" so if the engines in the selected category does not return any results more engines are queried? AFAIK @asciimoo wants to implement something similar, where searx first queries "priority" engines and then regular engines. The priority engines could be local ones so searx would not have to connect to any outside resource.

Could you please point me to the commits in your fork where you implemented this feature?

**Comment by wcerfgba** [comment]
Hi @kvch :wave: 

Sorry it's a bit hard to follow from my fork's history at the moment, there is a lot of sausage-making going on! The initial idea is contained in commit https://github.com/wcerfgba/searx/commit/fb12d9ff4458fc9ecbaa356039ae4041c20ce02f but I have revised this more since then to support multiple groups and to distribute the requests over multiple engines.

Effectively what this is doing is taking the user's initial query `burrito` and then building many new queries `burrito site:A`, `burrito site:B` , and then distributing those requests across all engines for which the feature is enabled. So it's not quite the same as having multiple engines, because then I would have to associate a site with an engine, i.e. build a "search site A on Google" engine.

You are correct that it could be implemented as engines, but I found it easier to get my head around building it out as query generation rather than adding new engines. Plus I think being able to distribute the load over multiple engines is useful, since this approach generates a lot of queries.

Happy to clean up my commit history and file a properly formatted PR for you this week if that makes it easier to see what's going on! :smiley:

**Comment by prp-e** [comment]
Hey @wcerfgba, hope you're doing well. 
Can you explain which part of the code you actually modified for that?

**Comment by wcerfgba** [comment]
> Hey @wcerfgba, hope you're doing well. Can you explain which part of the code you actually modified for that?

Hello @prp-e ! What specifically can I help you with? The most important changes are in https://github.com/wcerfgba/searx/commit/fb12d9ff4458fc9ecbaa356039ae4041c20ce02f but I can rebase and tidy up the subsequent history if that is useful for you :)
